### PR TITLE
implement globalplatform secure channel commands

### PR DESCRIPTION
**Pull request by gravityblast**
This is the first PR of 2 to import the installation process of the applet with globalplatform commands.
I started with 1 PR so that it's easier to review and it's usable even without the second part, at least to open the secure channel with the external authenticate command.

### CODE REVIEW

File: lib/src/main/java/im/status/hardwallet_lite_android/smartcard/Crypto.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        } catch (NoSuchAlgorithmException | NoSuchPaddingException e) {
+            throw new IllegalStateException("error generating session keys.", e);
+        } catch (InvalidKeyException | IllegalBlockSizeException | BadPaddingException | InvalidAlgorithmParameterException e) {
+            throw new RuntimeException("error generating session keys.", e);
+        }
+    }
+
+    public static byte[] appendDESPadding(byte[] data) {
+        int length = data.length + 1;
+        for (; length % 8 != 0; length++){
```
Comment (ligi) [inline lib/src/main/java/im/status/hardwallet_lite_android/smartcard/Crypto.java]: I think this:
```
  if (length % 8 != 0){
       length += 8-(length%8);
  }
```
would be a bit nicer for readability and efficiency

File: lib/src/main/java/im/status/hardwallet_lite_android/smartcard/Crypto.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+            Cipher cipher = Cipher.getInstance("DES/ECB/NoPadding");
+            SecretKeySpec key = new SecretKeySpec(resizeKey8(macKeyData), "DES");
+            cipher.init(Cipher.ENCRYPT_MODE, key);
+            return cipher.doFinal(mac);
+        } catch (GeneralSecurityException e) {
+            throw new RuntimeException("error generating ICV.", e);
+        }
+    }
+
+    public static byte[] generatePairingKey(char[] pairing) throws NoSuchAlgorithmException, InvalidKeySpecException {
```
Comment (bitgamma) [inline lib/src/main/java/im/status/hardwallet_lite_android/smartcard/Crypto.java]: This method is should not be needed anymore, it is integrated in the command set

File: lib/src/main/java/im/status/hardwallet_lite_android/smartcard/Crypto.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+import static android.util.Base64.NO_PADDING;
+
+public class Crypto {
+    public static final byte[] NullBytes8 = new byte[]{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};
+
+    public static long PIN_BOUND = 999999L;
+    public static long PUK_BOUND = 999999999999L;
+
+    public static byte[] deriveKey(byte[] cardKey, byte[] seq, byte[] purposeData) {
```
Comment (bitgamma) [inline lib/src/main/java/im/status/hardwallet_lite_android/smartcard/Crypto.java]: Let's make it clear that this scp02 session key derivation

File: lib/src/main/java/im/status/hardwallet_lite_android/smartcard/APDUWrapper.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
+package im.status.hardwallet_lite_android.smartcard;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.lang.reflect.Array;
+import java.util.Arrays;
+
+import im.status.hardwallet_lite_android.io.APDUCommand;
+
+public class APDUWrapper {
```
Comment (bitgamma) [inline lib/src/main/java/im/status/hardwallet_lite_android/smartcard/APDUWrapper.java]: Maybe SCP02Wrapper?

File: lib/src/main/java/im/status/hardwallet_lite_android/smartcard/SecureChannel.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    public APDUResponse send(APDUCommand cmd) throws IOException {
+        APDUCommand wrappedCommand = this.wrapper.wrap(cmd);
+        return this.channel.send(wrappedCommand);
+    }
+
+    public static Session verifyChallenge(byte[] hostChallenge, Keys cardKeys, APDUResponse resp) throws APDUException {
+        if (resp.getSw() == APDUResponse.SW_SECURITY_CONDITION_NOT_SATISFIED) {
+            throw new APDUException(resp.getSw(), "security confition not satisfied");
```
Comment (ligi) [inline lib/src/main/java/im/status/hardwallet_lite_android/smartcard/SecureChannel.java]: confition -> condition

File: lib/src/main/java/im/status/hardwallet_lite_android/smartcard/Keys.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
+package im.status.hardwallet_lite_android.smartcard;
+
+public class Keys {
```
Comment (bitgamma) [inline lib/src/main/java/im/status/hardwallet_lite_android/smartcard/Keys.java]: SCP02Keys?

File: lib/src/main/java/im/status/hardwallet_lite_android/smartcard/Crypto.java (hunk 1 above)
Comment (gravityblast) [reply lib/src/main/java/im/status/hardwallet_lite_android/smartcard/Crypto.java]: thank you @ligi, I don't know why it's still like that, it's an old part of the old project, good catch!

**Comment by gravityblast** [comment]
I can rename `package im.status.hardwallet_lite_android.smartcard` to `package im.status.hardwallet_lite_android.globalplatform`, would that be ok?

**Comment by bitgamma** [comment]
> I can rename `package im.status.hardwallet_lite_android.smartcard` to `package im.status.hardwallet_lite_android.globalplatform`, would that be ok?

Yes, but since we might be doing SCP03 later it is better to clarify what is SCP02 specific, what is GlobalPlatform specific, etc

File: lib/src/main/java/im/status/hardwallet_lite_android/smartcard/Crypto.java (hunk 1 above)
Comment (gravityblast) [reply lib/src/main/java/im/status/hardwallet_lite_android/smartcard/Crypto.java]: @ligi done! I removed the if because at least a `0x80` is always needed

**Comment by gravityblast** [comment]
@bitgamma I updated the code after the review. When the code is fine for you we can merge it and continue with the second part and maybe a third one for documentation

**Pull request merged**